File: frontend/src/api/violation.ts
// Layer 1: API calls — depends on types (Layer 0) only
import axios from 'axios'
import type { Violation, ApiResponse, PaginatedResult } from '@/types'

const client = axios.create({
  baseURL: '/api/v1',
  timeout: 10000,
})

// Add auth token to requests
client.interceptors.request.use((config) => {
  const token = localStorage.getItem('token')
  if (token) {
    config.headers.Authorization = `Bearer ${token}`
  }
  return config
})

export interface CreateViolationDto {
  student_id: string
  type: 'late_return' | 'noise' | 'damage' | 'property_violation' | 'other'
  description: string
  points: number
  handled_by: string
}

export interface UpdateViolationDto {
  type?: 'late_return' | 'noise' | 'damage' | 'property_violation' | 'other'
  description?: string
  points?: number
  handled_by?: string
  status?: 'pending' | 'resolved'
}

export interface ResolveViolationDto {
  status: 'resolved'
}

export interface ListViolationQuery {
  student_id?: string
  type?: string
  status?: string
  page?: number
  page_size?: number
}

export const violationApi = {
  list(query: ListViolationQuery = {}) {
    return client.get<ApiResponse<PaginatedResult<Violation>>>('/violations', {
      params: {
        student_id: query.student_id,
        type: query.type,
        status: query.status,
        page: query.page ?? 1,
        page_size: query.page_size ?? 20,
      },
    })
  },

  get(id: string) {
    return client.get<ApiResponse<Violation>>(`/violations/${id}`)
  },

  create(data: CreateViolationDto) {
    return client.post<ApiResponse<Violation>>('/violations', data)
  },

  update(id: string, data: UpdateViolationDto) {
    return client.put<ApiResponse<Violation>>(`/violations/${id}`, data)
  },

  delete(id: string) {
    return client.delete<ApiResponse<null>>(`/violations/${id}`)
  },

  resolve(id: string, data: ResolveViolationDto) {
    return client.put<ApiResponse<Violation>>(`/violations/${id}/resolve`, data)
  },
}

File: internal/request/violation.go
// Package request defines HTTP request DTOs.
// Layer 3: Depends on types (Layer 0). No business logic.
package request

import (
	"time"

	"github.com/example/dormitory-management/internal/types"
)

// CreateViolationRequest is the request body for creating a violation.
type CreateViolationRequest struct {
	StudentID   string `json:"student_id" binding:"required"`
	Type        string `json:"type" binding:"required,oneof=late_return noise damage property_violation other"`
	Description string `json:"description" binding:"required"`
	Points      int    `json:"points" binding:"required,min=1,max=100"`
	HandledBy   string `json:"handled_by" binding:"required"`
}

// ToViolation converts the request to a Violation type.
func (r *CreateViolationRequest) ToViolation() *types.Violation {
	now := time.Now()
	return &types.Violation{
		StudentID:   r.StudentID,
		Type:        r.Type,
		Description: r.Description,
		Points:      r.Points,
		HandledBy:   r.HandledBy,
		HandledAt:   now,
		Status:      "pending",
		CreatedAt:   now,
	}
}

// UpdateViolationRequest is the request body for updating a violation.
type UpdateViolationRequest struct {
	Type        *string `json:"type,omitempty"`
	Description *string `json:"description,omitempty"`
	Points      *int    `json:"points,omitempty"`
	HandledBy   *string `json:"handled_by,omitempty"`
	Status      *string `json:"status,omitempty" binding:"omitempty,oneof=pending resolved"`
}

// ApplyTo applies the update to a violation.
func (r *UpdateViolationRequest) ApplyTo(v *types.Violation) {
	if r.Type != nil {
		v.Type = *r.Type
	}
	if r.Description != nil {
		v.Description = *r.Description
	}
	if r.Points != nil {
		v.Points = *r.Points
	}
	if r.HandledBy != nil {
		v.HandledBy = *r.HandledBy
	}
	if r.Status != nil {
		v.Status = *r.Status
	}
}

// ResolveViolationRequest is the request body for resolving a violation.
type ResolveViolationRequest struct {
	Status string `json:"status" binding:"required,oneof=resolved"`
}

// ListViolationQuery holds query parameters for listing violations.
type ListViolationQuery struct {
	StudentID string `form:"student_id"`
	Type      string `form:"type"`
	Status    string `form:"status"`
	Page      int    `form:"page,default=1"`
	PageSize  int    `form:"page_size,default=20"`
}
